src/shell/doom.c

```diff
@@ -0,0 +1,243 @@
+        || flux_kvs_txn_put (txn,
+                             FLUX_KVS_APPEND,
+                             "exec.eventlog",
+                             entrystr) < 0
+        || !(f = flux_kvs_commit (doom->shell->h, NULL, 0, txn)))
+        shell_log_errno ("error posting task-exit eventlog entry");
+
+    if (f && doom->timeout != TIMEOUT_NONE)
+        flux_watcher_start (doom->timer);
+
+    free (entrystr);
+    flux_kvs_txn_destroy (txn);
+    flux_future_destroy (f); // fire and forget
+}
+
+static void doom_notify_cb (flux_t *h,
+                            flux_msg_handler_t *mh,
+                            const flux_msg_t *msg,
+                            void *arg)
+{
+    struct shell_doom *doom = arg;
+    json_t *entry;
+
+    assert (doom->shell->info->shell_rank == 0);
+
+    if (doom->done)
+        return;
+    if (flux_request_unpack (msg, NULL, "o", &entry) < 0) {
+        shell_log_errno ("error parsing first task exit notification");
+        return;
+    }
+    doom_post (doom, entry);
+    doom->done = true;
+}
+
+static void doom_notify (struct shell_doom *doom, json_t *entry)
+{
+    flux_future_t *f;
+
+    assert (doom->shell->info->shell_rank > 0);
+
+    if (!(f = flux_shell_rpc_pack (doom->shell,
+                                   "doom",
+                                   0,
+                                   FLUX_RPC_NORESPONSE,
+                                   "O",
+                                   entry)))
+        shell_log_errno ("error notifying rank 0 of first task exit");
+    flux_future_destroy (f);
+}
+
+static json_t *doom_eventlog_entry_create (flux_shell_task_t *task)
+{
+    json_t *task_info;
+    json_t *entry;
+
+    if (flux_shell_task_info_unpack (task, "o", &task_info) < 0
+        || !(entry = eventlog_entry_pack (0.,
+                                          "shell.task-exit",
+                                          "O",
+                                          task_info)))
+        return NULL;
+    return entry;
+}
+
+static void doom_timeout (flux_reactor_t *r,
+                          flux_watcher_t *w,
+                          int revents,
+                          void *arg)
+{
+    struct shell_doom *doom = arg;
+    char fsd[64];
+    fsd_format_duration (fsd, sizeof (fsd), doom->timeout);
+    shell_die (1, "%s timeout after first task exit", fsd);
+}
+
+static int doom_task_exit (flux_plugin_t *p,
+                           const char *topic,
+                           flux_plugin_arg_t *args,
+                           void *arg)
+{
+    flux_shell_t *shell;
+    struct shell_doom *doom;
+    flux_shell_task_t *task;
+
+    if (!(shell = flux_plugin_get_shell (p))
+        || !(doom = flux_plugin_aux_get (p, "doom"))
+        || !(task = flux_shell_current_task (shell)))
+        return -1;
+    if (!doom->done) {
+        json_t *entry;
+
+        if (!(entry = doom_eventlog_entry_create (task)))
```

Comment:
Hm, maybe it would be better to create the eventlog entry on the rank 0 shell? This avoids strange timestamps when clocks have drifted:

```ƒ(s=392,d=0,builddir) grondo@fluke2:~/git/f.git$ flux mini run -N 392 -n 392 -vvv sh -c 'sleep .$((392 - $FLUX_TASK_RANK))'
jobid: ƒ3s3MjkBZ
0.000s: job.submit {"userid":6885,"urgency":16,"flags":0}
0.014s: job.depend
0.014s: job.priority {"priority":16}
0.021s: job.alloc {"annotations":{"sched":{"resource_summary":"rank[0-391]/core0"}}}
0.382s: job.start
0.022s: exec.init
0.024s: exec.starting
0.327s: exec.shell.init {"leader-rank":0,"size":392,"service":"6885-shell-6320950345728"}
0.379s: exec.shell.start {"task-count":392}
-10.770s: exec.shell.task-exit {"localid":0,"rank":391,"state":"Exited","pid":1327786,"wait_status":0,"signaled":0,"exitcode":0}
1.335s: exec.complete {"status":0}
1.335s: exec.done
1.335s: job.finish {"status":0}
1.339s: job.release {"ranks":"all","final":true}
1.342s: job.free
1.342s: job.clean
```